test/benchmark_http_client_test.cc

```diff
@@ -41,41 +61,55 @@ class BenchmarkClientHttpTest : public Test {
 
     auto& tracer = static_cast<Envoy::Tracing::MockHttpTracer&>(*http_tracer_);
     EXPECT_CALL(tracer, startSpan_(_, _, _, _))
-        .WillRepeatedly(
-            Invoke([&](const Envoy::Tracing::Config& config, const Envoy::Http::HeaderMap&,
-                       const Envoy::StreamInfo::StreamInfo&,
-                       const Envoy::Tracing::Decision) -> Envoy::Tracing::Span* {
-              EXPECT_EQ(Envoy::Tracing::OperationName::Egress, config.operationName());
-              auto* span = new NiceMock<Envoy::Tracing::MockSpan>();
-              return span;
-            }));
-    request_generator_ = []() {
-      auto header = std::make_shared<Envoy::Http::TestRequestHeaderMapImpl>(
-          std::initializer_list<std::pair<std::string, std::string>>(
-              {{":scheme", "http"}, {":method", "GET"}, {":path", "/"}, {":host", "localhost"}}));
-      return std::make_unique<RequestImpl>(header);
-    };
+        .WillRepeatedly([&](const Envoy::Tracing::Config& config, const Envoy::Http::HeaderMap&,
+                            const Envoy::StreamInfo::StreamInfo&,
+                            const Envoy::Tracing::Decision) -> Envoy::Tracing::Span* {
+          EXPECT_EQ(Envoy::Tracing::OperationName::Egress, config.operationName());
+          auto* span = new NiceMock<Envoy::Tracing::MockSpan>();
+          return span;
+        });
   }
 
-  void testBasicFunctionality(const uint64_t max_pending, const uint64_t connection_limit,
-                              const uint64_t amount_of_request) {
+  // Used to set up benchmarkclient. Especially from within the testBenchmarkClientFunctionality.
+  void setupBenchmarkClient(const RequestGenerator& request_generator) {
+    client_ = std::make_unique<Client::BenchmarkClientHttpImpl>(
+        *api_, *dispatcher_, store_, std::make_unique<StreamingStatistic>(),
+        std::make_unique<StreamingStatistic>(), std::make_unique<StreamingStatistic>(),
+        std::make_unique<StreamingStatistic>(), false, cluster_manager_, http_tracer_, "benchmark",
+        request_generator, true);
+  }
+  // Primary testing method. Confirms that connection limits are met and number of requests are
+  // correct. If not empty, also checks the header expectations, if empty, it is ignored.
+  void TestBenchmarkClientProcessesExpectedInflightRequests(
+      const uint64_t max_pending, const uint64_t connection_limit, const uint64_t amount_of_request,
+      const RequestGenerator& request_generator,
+      const std::vector<absl::flat_hash_set<std::string>>& header_expectations =
+          std::vector<absl::flat_hash_set<std::string>>()) {
     if (client_ == nullptr) {
```

Comment:
Can we remove the if condition. The client_ state is not observable by the caller, and they might not understand why the generator lambda is not taking effect.

Reply:
I'm afraid that I don't really understand what this bit means. Also, not sure if it's still relevant now that I made it nullable.

Reply:
Now that I understand your comment, I agree with it, but because it affects other existing tests, I will be making that change in a separate PR to isolate it.

Reply:
let's file a bug instead for this.

Reply:
done